webhookValidator.go

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2021 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package ancla
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+)
+
+var (
+	errInvalidURL            = errors.New("Invalid Config URL")
+	errInvalidFailureURL     = errors.New("Invalid Failure URL")
+	errInvalidAlternativeURL = errors.New("Invalid Alternative URL(s)")
+	errURLIsNotHTTPS         = errors.New("URL is not HTTPS")
+	errInvalidHost           = errors.New("Invalid host")
+	errIPGivenAsHost         = errors.New("Cannot use IP as host")
+	errLocalHostGivenAsHost  = errors.New("Cannot use Localhost as host")
+	errLoopbackGivenAsHost   = errors.New("Cannot use loopback host")
+)
+
+type Validator interface {
+	Validate(w Webhook) error
+}
+
+type Validators []Validator
+
+type ValidFunc func(Webhook) error
+
+type ValidURLFunc func(*url.URL) error
+
+// Validate runs the given webhook through each validator in the validators list.
+// It returns as soon as the webhook is considered invalid and returns nil if the
+// webhook is valid.
+func (vs Validators) Validate(w Webhook) error {
+	for _, v := range vs {
+		err := v.Validate(w)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GoodURL parses the given webhook's Config.URL, FailureURL, and Config.AlternativeURLs
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodURL(v []ValidURLFunc) ValidFunc {
+	return func(w Webhook) error {
```

Comment:
Instead of checking our valid url functions aren't `nil` each time we try to use them, we can go through and fix our slice before returning our  `ValidFunc`. Above this `return` line, we can have a for loop that adds all non-nil `ValidURLFunc`s to a new slice.  Something like:

```go
fs := []ValidURLFunc{}
for _, v := range vs {
  if v != nil {
    fs = append(fs, v)
  }
}
```
And then use `fs` below, knowing they won't ever be nil.  The length of `fs` could be 0, but that's okay in this case.